pageserver/src/layered_repository/par_fsync.rs

```diff
@@ -0,0 +1,49 @@
+use std::{
+    fs::File,
+    io,
+    path::{Path, PathBuf},
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+fn fsync_path(path: &Path) -> io::Result<()> {
+    let file = File::open(path)?;
+    file.sync_all()
+}
+
+fn parallel_worker(paths: &[PathBuf], next_path_idx: &AtomicUsize) -> io::Result<()> {
+    while let Some(path) = paths.get(next_path_idx.fetch_add(1, Ordering::Relaxed)) {
+        fsync_path(path)?;
+    }
+
+    Ok(())
+}
+
+pub fn par_fsync(paths: &[PathBuf]) -> io::Result<()> {
+    const PARALLEL_PATH_THRESHOLD: usize = 1;
+    if paths.len() <= PARALLEL_PATH_THRESHOLD {
+        for path in paths {
+            fsync_path(path)?;
+        }
+        return Ok(());
+    }
+
+    const MAX_NUM_THREADS: usize = 64;
+    let num_threads = std::cmp::min(paths.len(), MAX_NUM_THREADS);
+    let next_path_idx = AtomicUsize::new(0);
+
+    crossbeam_utils::thread::scope(|s| -> io::Result<()> {
+        let mut handles = vec![];
+        for _ in 1..num_threads {
```

Comment:
A bunch of nits:

* Technically, that's `num_threads - 1` threads here, isn't it?
* `std::cmp::min(paths.len(), MAX_NUM_THREADS);` could be just `paths.len().min(MAX_NUM_THREADS);`

And a question out of curiosity:

* Is fixed `64` an adequate number to use?
One layer can easily spawn hundreds of files, so 64 threads might appear rather frequently, yet I wonder if my machine can normally handle that many tasks concurrently. I've read discussions before, that suggested a heuristics based on number of CPU cores available, but 64 is far from what I have either :)

Reply:
I had no answer to second question so I wrote a quick benchmark :) Also to compare fixed thread pool to spawning a thread per fsync call. It seems that because these are IO tasks they are not computationally intensive. So there is no need to limit the number of these threads to number of the available cpus. Actually tokio does the same, and by default number of IO threads is set to 512 (see [docs](https://docs.rs/tokio/1.13.0/tokio/runtime/struct.Builder.html#method.max_blocking_threads)). So it might be desirable to increase this limit. Technically this number is set for all spawn_blocking calls but tokio under the hood uses spawn_blocking for IO tasks.

Reply:
>     * Technically, that's `num_threads - 1` threads here, isn't it?

We spawn `num_threads - 1`, but the number of threads working is `num_threads`. I can clarify that in a comment.

>     * `std::cmp::min(paths.len(), MAX_NUM_THREADS);` could be just `paths.len().min(MAX_NUM_THREADS);`

Will do

>     * Is fixed `64` an adequate number to use?
>       One layer can easily spawn hundreds of files, so 64 threads might appear rather frequently, yet I wonder if my machine can normally handle that many tasks concurrently. I've read discussions before, that suggested a heuristics based on number of CPU cores available, but 64 is far from what I have either :)

As @LizardWizzard pointed out, these are not CPU intensive tasks. We could use `num_threads` >>> # of cores, if not for two tradeoffs:
- cost of memory
- cost of spawn/join latency

64 is somewhat arbitrary, I derived it from my tests in #722